packages/kbn-custom-events/src/index.ts

```diff
@@ -0,0 +1,180 @@
+      const [{ initializeFullStory }, userId] = await Promise.all([
+        fullStoryChunkPromise,
+        userIdPromise,
+      ]);
+
+      const { fullStory, sha256 } = initializeFullStory({
+        basePath,
+        orgId: fullstoryOrgId,
+        packageInfo,
+      });
+
+      // This needs to be called syncronously to be sure that we populate the user ID soon enough to make sessions merging
+      // across domains work
+      if (userId) {
+        // Join the cloud org id and the user to create a truly unique user id.
+        // The hashing here is to keep it at clear as possible in our source code that we do not send literal user IDs
+        const hashedId = sha256(`${esOrgId}:${userId}`);
+        const kibanaVer = packageInfo.version;
+        // TODO: use semver instead
+        const parsedVer = (kibanaVer.indexOf('.') > -1 ? kibanaVer.split('.') : []).map((s) =>
+          parseInt(s, 10)
+        );
+        fullStory.identify(
+          hashedId,
+          this.formatContext({
+            esOrgId,
+            version: kibanaVer,
+            versionMajor: parsedVer[0] ?? -1,
+            versionMinor: parsedVer[1] ?? -1,
+            versionPatch: parsedVer[2] ?? -1,
+          })
+        );
+        this.fullStory = fullStory;
+      }
+    } catch (e) {
+      // eslint-disable-next-line no-console
+      console.error(`[custom events] Could not call initialize due to error: ${e.toString()}`, e);
+    }
+
+    return !!this.fullStory;
+  }
+
+  private getFullstoryType(value: FullstoryEventValue) {
+    // For arrays, make the decidion based on the first element
+    const v = isArray(value) ? value[0] : value;
+    switch (typeof v) {
+      case 'string':
+        return 'str';
+      case 'number':
+        return isInteger(value) ? 'int' : 'real';
+      case 'boolean':
+        return 'bool';
+      case 'object':
+        if (v instanceof Date) {
+          return 'date';
+        }
+    }
+  }
+
+  private getCurrentMemoryState() {
+    // Get performance information from the browser (non standard property
+    // @ts-expect-error 2339
+    const memory = window.performance.memory;
+    let memInfo = {};
+    if (memory) {
+      memInfo = {
+        jsHeapSizeLimit: memory.jsHeapSizeLimit,
+        jsHeapSizeTotal: memory.totalJSHeapSize,
+        jsHeapSizeUsed: memory.usedJSHeapSize,
+      };
+    }
+    return memInfo;
+  }
+
+  private formatContext(context: Record<string, FullstoryEventValue>) {
+    // format context keys as required for env vars, see docs: https://help.fullstory.com/hc/en-us/articles/360020623234
+    return mapKeys(context, (value, key) => {
+      const snakeCasedKey = snakeCase(key);
+      const type = this.getFullstoryType(value);
+      return [snakeCasedKey, type].join('_');
+    });
+  }
+
+  public setUserContext(context: Record<string, FullstoryEventValue>) {
```

Comment:
are we going to use the same interface for the events on the server side?

Reply:
I don't see a reason why we wouldn't.

FYI FS does not support server side events atm, but of course another tool would.

Reply:
++ I understand that the idea behind this package is to provide a common interface and hide the actual underlying tools. So I guess that, internally, it will use different libs and tools to report the events depending on the environment UI vs. Server and the config. Is that correct?

Reply:
Yup, that's the point